test/integration/stats_integration_test.cc

```diff
@@ -129,5 +136,67 @@ TEST_P(StatsIntegrationTest, WithTagSpecifierWithFixedValue) {
   EXPECT_EQ(live->tags()[0].value_, "xxx");
 }
 
+// TODO(cmluciano) Refactor once https://github.com/envoyproxy/envoy/issues/5624 is solved
+// TODO(cmluciano) Add options to measure multiple workers & without stats
+// This class itself does not add additional tests. It is a helper for use in other tests measuring
+// cluster overhead.
+class ClusterMemoryTestHelper : public BaseIntegrationTest {
+public:
+  ClusterMemoryTestHelper()
+      : BaseIntegrationTest(testing::TestWithParam<Network::Address::IpVersion>::GetParam()) {}
+
+  /**
+   *
+   * @param num_clusters number of clusters appended to bootstrap_config
+   * @param allow_stats if false, enable set_reject_all in stats_config
+   * @return size_t the total memory allocated
+   */
+  size_t ClusterMemoryHelper(int num_clusters, bool allow_stats) {
+    config_helper_.addConfigModifier([&](envoy::config::bootstrap::v2::Bootstrap& bootstrap) {
+      if (!allow_stats) {
+        bootstrap.mutable_stats_config()->mutable_stats_matcher()->set_reject_all(true);
+      }
+      for (int i = 1; i < num_clusters; i++) {
+        auto* c = bootstrap.mutable_static_resources()->add_clusters();
+        c->set_name(fmt::format("cluster_{}", i));
+      }
+    });
+    initialize();
+
+    return Memory::Stats::totalCurrentlyAllocated();
+  }
+};
+class ClusterMemoryTestRunner : public testing::TestWithParam<Network::Address::IpVersion> {};
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, ClusterMemoryTestRunner,
+                         testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                         TestUtility::ipTestParamsToString);
+
+TEST_P(ClusterMemoryTestRunner, MemoryLargeClusterSizeWithStats) {
+  // Skip test if we cannot measure memory with TCMALLOC
+  if (!Stats::TestUtil::hasDeterministicMallocStats()) {
+    return;
+  }
+  const size_t start_mem = Memory::Stats::totalCurrentlyAllocated();
+  // A unique instance of ClusterMemoryTest allows for multiple runs of Envoy with
+  // differing configuration. This is necessary for measuring the memory consumption
+  // between the different instances within the same test.
+  auto t1 = std::make_unique<ClusterMemoryTestHelper>();
+  size_t m1 = t1->ClusterMemoryHelper(1, true);
+  EXPECT_LT(start_mem, m1);
+  t1.reset(nullptr);
+
+  auto t2 = std::make_unique<ClusterMemoryTestHelper>();
+  const size_t m1001 = t2->ClusterMemoryHelper(1001, true);
+  EXPECT_LT(start_mem, m1001);
+  size_t m_per_cluster = (m1001 - m1) / 1000;
+// As of 2019/03/13, m_per_cluster = 56404 (libstdc++), 52249 (libc++).
+#ifdef _LIBCPP_VERSION
+  EXPECT_LT(m_per_cluster, 53000);
+#else
+  EXPECT_LT(m_per_cluster, 57000);
```

Comment:
@jmarantz shouldn't we also add the restriction that this only applies to certain build types? To detect meaningful regression and make meaningful statement of memory use, we should only measure/check for release. Otherwise, things like changes to how ASAN or MSAN operate, in terms of padding and fencing, might impact these numbers.

Reply:
That's what the Stats::TestUtil::hasDeterministicMallocStats() check above is for.

Reply:
Hmm, I think that's a bad name, you can have deterministic stats which are still not reflective of production memory usage. E.g. what if you have different levels of padding in debug vs. opt builds?

Could we just limit this to a release build with `#ifdef`?

Reply:
I don't love limiting to release-builds as that makes it harder to debug memory problems :)

However I agree with you that the name is bad, and I think actually the methodology needs to grow up too. How about this idea; let's change Stats::TestUtil::hasDeterministicMallocStats()  to Stats::TestUtil::memoryProfile() which returns an3-choice enum consisting of the 2 current models that @cmluciano has sorted out in this PR, and an 'unknown' entry. We should switch on that enum to adjust the expected results for any particular test.

Also with that, we should document how to repro those testcases.

However I suggest we do that in a follow-up, because as we iterate over how to do this nicely, further bloats in other PRs build up unnoticed.

Reply:
SGTM